tests/61push/08_rejected_pushers.pl

```diff
@@ -0,0 +1,104 @@
+         kind                => "http",
+         app_id              => $app_id,
+         app_display_name    => "sytest_display_name",
+         device_display_name => "device_display_name",
+         pushkey             => $push_key,
+         lang                => "en",
+         data                => { url => $url },
+      },
+   );
+}
+
+sub wait_for_push
+{
+   my ( $path, $response ) = @_;
+   await_http_request( $path, sub {
+     my ( $request ) = @_;
+     my $body = $request->body_from_json;
+
+     return unless $body->{notification}{type};
+     return unless $body->{notification}{type} eq "m.room.message";
+     return 1;
+   })->then( sub {
+      my ( $request ) = @_;
+
+      $request->respond_json( $response // {} );
+      Future->done( $request );
+   });
+}
+
+multi_test "Test that rejected pushers are removed.",
+   requires => [
+      local_user_fixtures( 2, with_events => 0 ),
+      $main::TEST_SERVER_INFO,
+   ],
+
+   do => sub {
+      my ( $alice, $bob, $test_server_info ) = @_;
+
+      my $room_id;
+
+      my $url = $test_server_info->client_location . "/alice_push";
+
+      matrix_create_room( $alice, visibility => "private" )->then( sub {
+         ( $room_id ) = @_;
+         matrix_invite_user_to_room( $alice, $bob, $room_id ),
+      })->then( sub {
+         matrix_join_room( $bob, $room_id);
+      })->then( sub {
+         matrix_send_room_text_message(
+            $bob, $room_id, body => "message"
+         )->then( sub {
+            my ( $event_id ) = @_;
+            matrix_advance_room_receipt( $alice, $room_id,
+                "m.read" => $event_id
+            );
+         });
+      })->then( sub {
+         create_pusher( $alice, "sytest", "key_1", "$url/1" )
+            ->SyTest::pass_on_done( "Alice's pusher 1 created" );
+      })->then( sub {
+         create_pusher( $alice, "sytest", "key_2", "$url/2" )
+            ->SyTest::pass_on_done( "Alice's pusher 2 created" );
+      })->then( sub {
+         do_request_json_for( $alice,
+              method  => "GET",
+              uri     => "/r0/pushers",
+         )->then( sub {
+            my ( $body ) = @_;
+
+            assert_json_keys( $body, qw( pushers ) );
+            @{ $body->{pushers} } == 2 or die "Expected two pushers";
+
+            Future->done(1);
+         });
+      })->then( sub {
+         Future->needs_all(
+            wait_for_push( "/alice_push/1",  { rejected => [ "key_1" ] } ),
+            wait_for_push( "/alice_push/2" ),
+            matrix_send_room_text_message( $bob, $room_id, body => "message" )
+               ->SyTest::pass_on_done( "Message Sent" ),
+         );
+      })->then( sub {
+         do_request_json_for( $alice,
+              method  => "GET",
+              uri     => "/r0/pushers",
+         )->then( sub {
+            my ( $body ) = @_;
+
+            assert_json_keys( $body, qw( pushers ) );
+            @{ $body->{pushers} } == 1 or die "Expected one pusher";
```

Comment:
For extra points, might you want to check that of the single pusher that got retained, it retained the correct one?